src/vulkan/descriptor.h

```diff
@@ -43,10 +43,15 @@ enum class DescriptorType : uint8_t {
 VkDescriptorType ToVkDescriptorType(DescriptorType type);
 
 struct BufferData {
-  DataType type;
   uint32_t offset;
   size_t size_in_bytes;
-  std::vector<Value> values;
+  DataType type;                      // Primitive type of |values|.
+  std::vector<Value> values;          // Data that has primitive
```

Comment:
What is a 'primitive type' ?Why don't we just have Value objects everywhere?

Reply:
Changed 'primitive type' to type.
`values` will be used for data given by script while `raw_data` will be used for read back from Vulkan buffer.
Since `Value` class is specialized for a value with type, it is not suitable to keep the raw data from Vulkan buffer.

Reply:
But, we created the vulkan buffer, so we should already know the type of the buffer we're reading from yea? So, we can create value objects of the correct type based off the buffer?

Reply:
What if user overwrite a single ssbo with multiple types of values?
```
ssbo 0:0 subdata float 0 1 2 3
ssbo 0:0 subdata int   4 4 5 6
```
In the above example, how do we determine the type of ssbo at offset 4? `int` or `float`?

I doubt this approach because Vulkan buffers e.g., ssbo and ubo do not have types and they are just memory chunk. I think the meaning of value in those buffers must be interpreted by users when probing them. In Vulkan engine layer, I think we should keep it as raw data.

Reply:
If you think we should not put `raw_data` into `struct BufferData`, we can separate them and keep `raw_data` in `Descriptor` class.

I also prefer this now. Does it sound good to you?

Reply:
In that case, you can just read all of the numbers out as a float into the Value object and cast at the caller side if you want something different.

Reply:
The main benefit I see of using Values is you only need to store one thing, instead of it being values, or raw if values is empty which I can see causing issues down the line. Just having a single storage place makes it easier to understand what's happening.

Reply:
Since we don't know what is the meaning of GPU memory when we read it back to CPU memory, we would better keep it as a memory chunk without type. Let me split `struct BufferData` into `struct BufferInput` and `struct BufferOutput`. `struct BufferInput` must be used for data given by user and `struct BufferOutput` must be used for data from GPU memory.